python/ray/data/read_api.py

```diff
@@ -2727,26 +2727,37 @@ def from_huggingface(
         A :class:`~ray.data.Dataset` holding rows from the `Hugging Face Datasets Dataset`_.
     """  # noqa: E501
     import datasets
+    from aiohttp.client_exceptions import ClientResponseError
 
     from ray.data.datasource.huggingface_datasource import HuggingFaceDatasource
 
     if isinstance(dataset, (datasets.IterableDataset, datasets.Dataset)):
-        # Attempt to read data via Hugging Face Hub parquet files. If the
-        # returned list of files is empty, attempt read via other methods.
-        file_urls = HuggingFaceDatasource.list_parquet_urls_from_dataset(dataset)
-        if len(file_urls) > 0:
-            # If file urls are returned, the parquet files are available via API
-            # TODO: Add support for reading from http filesystem in FileBasedDatasource
-            # GH Issue: https://github.com/ray-project/ray/issues/42706
-            import fsspec.implementations.http
-
-            http = fsspec.implementations.http.HTTPFileSystem()
-            return read_parquet(
-                file_urls,
-                parallelism=parallelism,
-                filesystem=http,
-                concurrency=concurrency,
-                override_num_blocks=override_num_blocks,
+        try:
+            # Attempt to read data via Hugging Face Hub parquet files. If the
+            # returned list of files is empty, attempt read via other methods.
+            file_urls = HuggingFaceDatasource.list_parquet_urls_from_dataset(dataset)
+            if len(file_urls) > 0:
+                # If file urls are returned, the parquet files are available via API
+                # TODO: Add support for reading from http filesystem in
+                # FileBasedDatasource. GH Issue:
+                # https://github.com/ray-project/ray/issues/42706
+                import fsspec.implementations.http
+
+                http = fsspec.implementations.http.HTTPFileSystem()
+                return read_parquet(
+                    file_urls,
+                    parallelism=parallelism,
+                    filesystem=http,
+                    concurrency=concurrency,
+                    override_num_blocks=override_num_blocks,
+                    ray_remote_args={
+                        "retry_exceptions": [FileNotFoundError, ClientResponseError]
```

Comment:
If a file is returned by `list_parquet_urls_from_dataset`, is that file guaranteed to exist? Wondering if Ray might indefinitely retry a file that's returned from `list_parquet_urls_from_dataset` but doesn't actually exist

Reply:
The default number of retries should just be 3 (from [here](https://docs.ray.io/en/latest/ray-core/api/doc/ray.remote.html): "max_retries – Only for remote functions. This specifies the maximum number of times that the remote function should be rerun when the worker process executing it crashes unexpectedly. The minimum valid value is 0, the default value is 3, and a value of -1 indicates infinite retries.).

The list of files returned should all exist since HuggingFace is returning a list of files they host, but it will try each file at most three times (slower than just one try but will terminate).

Reply:
If we use `cached_remote_fn`, we override the default with -1:
https://github.com/ray-project/ray/blob/5e2a27630346b4bbc19c336dc90affac6a8da24a/python/ray/data/_internal/remote_fn.py#L19-L28

> The list of files returned should all exist since HuggingFace is returning a list of files they host

Ah, should be fine then.